### DISCUSSION TITLE
Loading height maps and textures?

### DISCUSSION THREAD

**Discussion by niklasking** [spimort/TerraBrush#72] [Q&A] [answered]
This add-on is brilliant and most probably the best add-on I have ever seen for Godot! Well done!

I am trying to build a terrain based on real world data, where I have a height map for the elevation, but also shape files for different types of terrain. For instance, one shape file for forest, one for fields and one for water. 
From what I understand from your zone concept, this information should fit perfectly in your resource files, if I just can figure out how to make it happen :-)

So, any idea how I can convert a red/black image to the resource file you are using for the zones?

Best regards and thanks for the good work!
/Niklas

**Comment by spimort** [comment, marked as answer]
Hey @niklasking !

I'm so glad you enjoy what I've done!

So, the whole system is based on heightmaps, as long as you have ImageTexture resource in Godot, you should be able to use it in the zones.
For the heightmap itself, I use the RF format (OpenGL texture format GL_R32F where there's one component, a 32-bit floating-point value.) so I can have somewhat precise values that goes over 255.

I made a quick test here, I took a heightmap from new york (only the area with the squares) : 
![image](https://github.com/spimort/TerraBrush/assets/6035485/3ab8159b-4652-468f-9f89-5d6e0116c083)

and I made a script to convert the resulting image into 4 zones : 
![image](https://github.com/spimort/TerraBrush/assets/6035485/0c3bf6de-dc9f-4183-9eb0-0d82a21cb436)
![image](https://github.com/spimort/TerraBrush/assets/6035485/87d9b46c-441e-4e0c-b98c-5b1b4437a754)
![image](https://github.com/spimort/TerraBrush/assets/6035485/9c500d35-1808-4831-946c-2414cf320e8a)

Obviously my source image is not really precise but that gives you a picture of what could be done.

This was just a quick test, but here's what I did to generate the zones images : 
```cs
    public void Test() {
        var heightmap = ResourceLoader.Load<Texture2D>("res://Tester/heightmap.png");
        var image = heightmap.GetImage();

        float factor = 10.0f; // Just a simple factor to increase height
        // Generate 4 zones from the original image (the original image is 1080x1080 so each zones will be of 540)
        /// The -1.3f is the minium height for this zone and the 106f is the maximum height for this zone
        GenerateImageTextureForZone(image, 0, 0, 540, -1.3f, 106f, factor);
        GenerateImageTextureForZone(image, 1, 0, 540, -1.3f, 106f, factor);
        GenerateImageTextureForZone(image, 0, 1, 540, -1.3f, 106f, factor);
        GenerateImageTextureForZone(image, 1, 1, 540, -1.3f, 106f, factor);
    }

    private void GenerateImageTextureForZone(Image image, int zoneX, int zoneY, int size, float min, float max, float factor) {
        var newImage = Image.Create(size, size, false, Image.Format.Rf); // Create a RF image

        var startingX = size * zoneX;
        var startingY = size * zoneY;
        for (var x = 0; x < size; x++) {
            for (var y = 0; y < size; y++) {
                var toX = x + startingX;
                var toY = y + startingY;
                var pixel = Colors.Black;

                try {
                    pixel = image.GetPixel(toX, toY);
                } catch {
                    // Throw an actual exception, this was just for testing
                }

                var heightValue = Mathf.Remap(pixel.R, 0f, 1f, min, max) * factor;
                newImage.SetPixel(x, y, new Color(heightValue, 0, 0, 1f)); // Set the red channel value (we are on a RF image so only the red channel is relevant)
            }
        }

        var imageTexture = ImageTexture.CreateFromImage(newImage);
        ResourceSaver.Save(imageTexture, $"res://Tester/h_{zoneX}_{zoneY}.res"); // Save the image as a resource so we can use it in the zones.
    }
```

Does that make sense to you ? 🙂

**Reply by niklasking** [reply]
This is brilliant and works perfectly! Thanks a lot! So simple :-)

Next question is how to create the foliage, object, water and splatmap file/images. What values shall I use to get the correct texture, object, ...?

Best regards,
Niklas

**Reply by spimort** [reply]
It's all the same kind of stuff, create an ImageTexture resource with the images : 

- **Textures** : It uses splatmaps. You can have several splatmaps, each of them are RGBA8 and each channel correspond to a texture. It means that each splatmap can store the information of maximum 4 textures. If you have more than 4 textures, you need more splatmaps.
- **Foliage** : It uses a RGBA8 image. As of now, only the R channel is being used. The R value will correspond to the size of the foliage. The maximum being 1.0 will correspond to the maximum size of the foliage (useful to create transitions between grass and no grass)
- **Objects** : It uses a RGBA8 image. As of now only the R  channel is being used. The R value only means "Present" otherwise, no object will display.
- **Water** : It uses a RGBA8 image. The R channel correspond to the presence of the water and also the deepness of it (the maximum is 1.0 and is multiplied by the water factor). The G and B channel are used to store the flow of the water, to give it movement.
- **Snow** : It uses a RGBA8 image, As of now, only the R channel is being used. The R value correspond to how much snow there is (the maximum is 1.0 and is multiplied by the snow factor)

I hope this helps! 🙂

**Reply by niklasking** [reply]
I just added two textures for four zones (625x625m) and it works perfectly fine.
Thanks, thanks a lot!

Best regards from a super happy Niklas :-)

**Reply by niklasking** [reply]
One more comment:

Generating foliage and objects also works fine. However, for objects it looks like all generated objects are always shown, not as for the foliage where you hide things that are "far" away. This means that in my forest of 2 500 km x 2 500 km there will be a lot of trees and my computer cannot handle this.
Are you planning to implement the same behavior for objects as for foliage? Or have I misunderstood something?

Except for that, it all works perfectly fine and I am still impresser :-)

Best regards,
Niklas

**Reply by spimort** [reply]
Hey @niklasking !

The objects are not done the same way, but godot already handles that for us!
When you create a packedscene with your model, you can set the visibility range of the mesh : 
![image](https://github.com/spimort/TerraBrush/assets/6035485/20025903-4a3a-469a-8196-2262cde9d2d0)

That should make it!

**Reply by niklasking** [reply]
Thanks a lot!
That I did not know, but it works fine.

However, and sorry to trouble you, but something is strange when I generate trees (objects). It looks like the texture map/image is not considered and also, it looks like the object frequency is also not considered.
I created a small terrain and used the same image for texture, foliage and for objects, just to make sure the image was not wrong. The texture and the foliage works fine, but not the objects. Am I missing something?
![image](https://github.com/spimort/TerraBrush/assets/29523500/cdd6a2dc-a6d9-4c3a-b188-65cc26482f65)
The object frequency is set to 50 m, but whatever value I use I get the same. And as you can see, the texture map is not used.
![image](https://github.com/spimort/TerraBrush/assets/29523500/c1a81015-3a2c-4531-956b-c46ccd9e4632)

**Reply by spimort** [reply]
No worries, I'm always happy to help!

For the object frequency, this property is used while painting, to determine how often we paint on the terrain. So it wont affect the objects when you place an image that is pre-generated.

For the black spot you have there, from your splatmaps, it seems like you have at least 4 textures, because white would mean that you set 1.0 to all 4 channels of the image. If so, the splatmaps is telling the system to try to blend all 4 textures together, but they all have the same amount. I could be wrong but this is what I think from what I see here. Maybe you could try to paint on the terrain with the tool and see what kind of image it produces, it might help you 🙂.

Your grass texture seems to be very pixelated, maybe try to generate the mipmaps for your textures (in the import panel, when selecting the image).

I hope this helps!

**Reply by niklasking** [reply]
Hi and thanks!
Checking the generated mipmaps was a brilliant idea :-) It looks like the mipmap is an RGBA8 with (1, 0, 0, 1) for trees and (0, 0, 0, 0) for no tree.
I divided my 2496x2496m terrain in 64 zones (312x312m) and it works fine for the first zone:
![image](https://github.com/spimort/TerraBrush/assets/29523500/588c5778-e43c-4e11-b146-d84481be17b2)
However, for the next zones, the trees (objects) are not shown and I get the following error when I reload the project. Any ideas?

/root/godot/modules/mono/glue/GodotSharp/GodotSharp/Core/NativeInterop/ExceptionUtils.cs:112 - System.NullReferenceException: Object reference not set to an instance of an object.
     at TerraBrush.CustomContentLoader.AddObjectsPreviewToParent(TerraBrush terraBrush, Node parentNode, Action`1 onSelect, Boolean useCircleIcon) in C:\Users\nikla\OneDrive\Dokument\godot-projekts\Gronadal-20240419\addons\terrabrush\Scripts\CustomContentLoader.cs:line 111
     at TerraBrush.TerrainControlDock.InitializeObjects() in C:\Users\nikla\OneDrive\Dokument\godot-projekts\Gronadal-20240419\addons\terrabrush\Scripts\TerrainControlDock.cs:line 183
     at TerraBrush.TerrainControlDock._Ready() in C:\Users\nikla\OneDrive\Dokument\godot-projekts\Gronadal-20240419\addons\terrabrush\Scripts\TerrainControlDock.cs:line 44
     at Godot.Node.InvokeGodotClassMethod(godot_string_name& method, NativeVariantPtrArgs args, godot_variant& ret) in /root/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/Node.cs:line 2117
     at Godot.CanvasItem.InvokeGodotClassMethod(godot_string_name& method, NativeVariantPtrArgs args, godot_variant& ret) in /root/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/CanvasItem.cs:line 1370
     at Godot.Control.InvokeGodotClassMethod(godot_string_name& method, NativeVariantPtrArgs args, godot_variant& ret) in /root/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/Control.cs:line 2841
     at TerraBrush.TerrainControlDock.InvokeGodotClassMethod(godot_string_name& method, NativeVariantPtrArgs args, godot_variant& ret) in C:\Users\nikla\OneDrive\Dokument\godot-projekts\Gronadal-20240419\Godot.SourceGenerators\Godot.SourceGenerators.ScriptMethodsGenerator\TerraBrush.TerrainControlDock_ScriptMethods.generated.cs:line 230
     at Godot.Bridge.CSharpInstanceBridge.Call(IntPtr godotObjectGCHandle, godot_string_name* method, godot_variant** args, Int32 argCount, godot_variant_call_error* refCallError, godot_variant* ret) in /root/godot/modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/CSharpInstanceBridge.cs:line 24

Finally, is there any way to place the trees (objects) more separated even if I use my own map? The forest is right now very dense :-)

I am so close now and once again thanks for the very good add-on and the brilliant support.

Best regards,
Niklas

**Reply by spimort** [reply]
Hum, that's really hard to tell without seing the project. Would you be able to share it with me (feel free to remove "sensitive" stuff, as long as I can reproduce the problem)? If so, you could upload it on something like "wetransfer". If needed, feel free to join my discord server, it might be easier to share it there : https://discord.com/invite/htktb7k

For the density right now there is no way to do it without painting scape between the objects but this is something that could be done! Could you open an issue for that ? 🙂

**Comment by spimort** [comment]
Hey @niklasking !
I know it's been a long time, but the version 0.10.0Alpha now supports the import/export 😎
